### ISSUE TITLE
no-extra-parens false positive with objects that have their keys read out immediately

### ISSUE THREAD

**Issue by Famlam** [eslint/eslint#4427]
With option `no-extra-parens: 1` the following JS code:
```javascript
({})["hasOwnProperty"]("hasOwnProperty");
```
generates the warning `Gratuitous parentheses around expression.`

This can be "resolved" by using 
```javascript
{}["hasOwnProperty"]("hasOwnProperty");
```

However, the latter generates a syntax error (at least in Chrome), while the former is valid JS (and returns `false`)

I'm using the current version on the web demo (presumably 1.9.0)

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by michaelficarra** [comment]
The entire expression should be parenthesised, since that's all that is strictly necessary. The rule is correct here.

```js
({}["hasOwnProperty"]("hasOwnProperty"));
```

**Comment by nzakas** [comment]
@Famlam do you have a larger example where this is used? As-is, the statement does nothing.

**Comment by Famlam** [comment]
@nzakas yes, I encountered it in the following code:
```javascript
var isObjectProperty = function(str) {
  return ({})[str] !== undefined;
};
```

**Comment by michaelficarra** [comment]
You've moved your expression to statement position, which changes things. Leave it in expression position and you're fine. Otherwise, you're allowed to parenthesise the whole ExpressionStatement. There's no bug here, really.

**Comment by nzakas** [comment]
Ah, good point. When used with `return`, those parens are not required. Thanks @michaelficarra
